File: java_vs_javascript/benchmark_tests/BenchmarkTests.java
package java_vs_javascript.benchmark_tests;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class BenchmarkTests {

    /**
     * Java vs Javascript Benchmark Test
     * - replace the function you want to run
     * 
     * @param args
     * @throws IOException
     */
    public static void main(String[] args) throws IOException {
        ArraySort();
    }

    /**
     * JAVA wins
     * sum: 499999999500000000
     * Time: 353.422166 ms
     */
    public static void loopSample() {
        System.out.println("\n==> Loop Sample");
        long start = System.nanoTime();
        long sum = 0;
        for (long i = 0; i < 1_000_000_000; i++) {
            sum += i;
        }
        long end = System.nanoTime();
        System.out.println("sum: " + sum);
        System.out.println("Time: " + (end - start) / 1e6 + " ms");
    }

    /**
     * JAVA wins
     * Sum: 49999995000000
     * Time: 265.91675 ms
     */
    public static void memoryIntensiveSample() {
        System.out.println("\n==> Memory Intensive Sample");
        long start = System.nanoTime();
        List<Integer> numbers = new ArrayList<>();
        for (int i = 0; i < 10_000_000; i++) {
            numbers.add(i);
        }
        long sum = numbers.stream().mapToLong(Integer::longValue).sum();
        long end = System.nanoTime();
        System.out.println("Sum: " + sum);
        System.out.println("Time: " + (end - start) / 1e6 + " ms");
    }

    /**
     * Javascript wins
     * Time: 169.921708 ms
     */
    public static void IOIntensiveSample() throws IOException {
        System.out.println("\n==> IO Intensive Sample");
        File file = new File("largeFileJava.txt");
        long start = System.nanoTime();

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
            for (int i = 0; i < 1_000_000; i++) {
                writer.write("This is a benchmark test.\n");
            }
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            while (reader.readLine() != null) {
            }
        }

        long end = System.nanoTime();
        System.out.println("Time: " + (end - start) / 1e6 + " ms");

    }

    /**
     * Java wins
     * 3.380916 ms
     */
    public static void StringConcat() {
        System.out.println("\n==> String Concat Test");
        StringBuilder str = new StringBuilder();
        long startTime = System.nanoTime();
        for (int i = 0; i < 1e5; i++) {
            str.append("a");
        }
        long endTime = System.nanoTime();
        System.out.println("StringConcat Java: " + (endTime - startTime) / 1e6 + " ms");
    }

    /**
     * 97.492125 ms
     */
    public static void ArraySort() {
        System.out.println("\n==> Array Sorting");
        double[] array = new Random().doubles(1_000_000).toArray();
        long startTime = System.nanoTime();
        Arrays.sort(array);
        long endTime = System.nanoTime();
        System.out.println("Sort Java: " + (endTime - startTime) / 1e6 + " ms");
    }

}

File: java_vs_javascript/benchmark_tests/BenchmarkTests.js
// @ts-check
const fs = require('fs');
const path = require('path');

class BenchmarkTests {
	/**
	 * Time: 1227 ms
	 */
	loopSample() {
		console.log('\n==> Loop Sample');
		const start = Date.now();
		let sum = 0;
		for (let i = 0; i < 1_000_000_000; i++) {
			sum += i;
		}
		const end = Date.now();
		console.log(`Time: ${end - start} ms`);
	}

	/**
	 * Sum: 49999995000000
	 * Time: 404 ms
	 */
	memoryIntensiveSample() {
		console.log('\n==> Memory Intensive Sample');
		const start = Date.now();
		const numbers = [];
		for (let i = 0; i < 10_000_000; i++) {
			numbers.push(i);
		}
		const sum = numbers.reduce((acc, num) => acc + num, 0);
		const end = Date.now();
		console.log('Sum:', sum);
		console.log('Time:', end - start, 'ms');
	}

	/**
	 * Time: 99ms
	 */
	IOIntensiveTask() {
		console.log('\n==> IO Intensive Sample');
		const filePath = path.join(__dirname, 'largeFileJS.txt');
		const start = Date.now();
		fs.writeFileSync(filePath, 'This is a benchmark test.\n'.repeat(1_000_000));
		const content = fs.readFileSync(filePath, 'utf-8');
		const end = Date.now();
		console.log('Time: ', end - start, ' ms');
	}

	/**
	 * 9.902 ms
	 */
	stringConcat() {
		console.time('stringConcat');
		let str = '';
		for (let i = 0; i < 1e5; i++) {
			str += 'a';
		}
		console.timeEnd('stringConcat');
	}

	/**
	 * 491.172ms
	 */
	arraySort() {
		console.time('SortJS');
		const array = Array.from({ length: 1e6 }, () => Math.random());
		array.sort((a, b) => b - a);
		console.timeEnd('SortJS');
	}
}

const benchmarkTests = new BenchmarkTests();
